src/main/java/at/hannibal2/skyhanni/data/PetAPI.kt

```
@@ -35,31 +194,412 @@ object PetAPI {
+            }
+
+            val tempPetMisc = handleWidgetPetLine(line, newPetLine)
+            if (tempPetMisc != null) {
+                newPetData = newPetData.copy(
+                    internalName = tempPetMisc.internalName,
+                    cleanName = tempPetMisc.cleanName,
+                    rarity = tempPetMisc.rarity,
+                    level = tempPetMisc.level,
+                    xp = tempPetMisc.xp,
+                    rawPetName = tempPetMisc.rawPetName,
+                )
+                return@forEach
+            }
+        }
+        fireEvent(newPetData.copy(xp = newPetData.xp + overflowXP))
+    }
+
+    private fun handleWidgetPetLine(line: String, newPetLine: String): PetData? {
+        petWidgetPattern.matchMatcher(line) {
+            val rarity = LorenzRarity.getByColorCode(group("rarity")[0]) ?: LorenzRarity.ULTIMATE
+            val petName = group("name")
+            val level = group("level").toInt()
+            val xp = levelToXP(level, rarity, petName) ?: return null
+
+            return PetData(
+                petNameToInternalName(petName, rarity),
+                petName,
+                rarity,
+                null,
+                level,
+                xp,
+                newPetLine,
+            )
+        }
+        return null
+    }
+
+    private fun handleWidgetStringLine(line: String): NEUInternalName? {
+        widgetStringPattern.matchMatcher(line) {
+            val string = group("string")
+            if (string == "No pet selected") {
+                PetChangeEvent(pet, null).post()
+                pet = null
+                return null
+            }
+            return NEUInternalName.fromItemNameOrNull(string)
+        }
+        return null
+    }
+
+    private fun handleWidgetXPLine(line: String): Double? {
+        xpWidgetPattern.matchMatcher(line) {
+            if (group("max") != null) return null
+
+            val overflow = group("overflow")?.formatDoubleOrNull() ?: 0.0
```

dont use formatDoubleOrNull, use formatDouble(), that will cause an error. if the format is wrong, we want to know about it, not hide the error